tree/ntuple/src/RField.cxx

```diff
@@ -1010,22 +1027,38 @@ std::size_t ROOT::ROptionalField::AppendImpl(const void *from)
    }
 }
 
-void ROOT::ROptionalField::ReadGlobalImpl(ROOT::NTupleSize_t globalIndex, void *to)
+void ROOT::ROptionalField::PrepareRead(void *to, bool hasOnDiskValue)
 {
    auto engagementPtr = GetEngagementPtr(to);
-   auto itemIndex = GetItemIndex(globalIndex);
-   if (itemIndex.GetIndexInCluster() == ROOT::kInvalidNTupleIndex) {
-      if (*engagementPtr && !(fSubfields[0]->GetTraits() & kTraitTriviallyDestructible))
-         fItemDeleter->operator()(to, true /* dtorOnly */);
-      *engagementPtr = false;
-   } else {
+   if (hasOnDiskValue) {
       if (!(*engagementPtr) && !(fSubfields[0]->GetTraits() & kTraitTriviallyConstructible))
          CallConstructValueOn(*fSubfields[0], to);
-      CallReadOn(*fSubfields[0], itemIndex, to);
       *engagementPtr = true;
+   } else {
+      if (*engagementPtr && !(fSubfields[0]->GetTraits() & kTraitTriviallyDestructible))
+         fItemDeleter->operator()(to, true /* dtorOnly */);
+      *engagementPtr = false;
    }
 }
 
+void ROOT::ROptionalField::ReadGlobalImpl(ROOT::NTupleSize_t globalIndex, void *to)
+{
+   auto itemIndex = GetItemIndex(globalIndex);
+   const bool hasOnDiskValue = itemIndex.GetIndexInCluster() != ROOT::kInvalidNTupleIndex;
+   PrepareRead(to, hasOnDiskValue);
```

Comment:
A thought for the future (doesn't need to be addressed in this PR): We may also want to return a pointer from `ROptionalField::PrepareRead` because I don't think the C++ standard guarantees that the contained value is the first member of `std::optional` nor that it necessarily is at the same address (for example with(out) empty base class optimization).